src/mvba/abba/mod.rs

```diff
@@ -1,56 +1,567 @@
+                    }
+                }
+            }
+
+            Action::PreVote(action) => {
+                if action.round != self.r {
+                    return Ok(());
+                }
+                // Let's check pre-votes:
+
+                // HECK FOR DECISION. Collect n −t valid and properly justified main-votes of round r .
+                // If these are all main-votes for b ∈ {0, 1}, then decide the value b for ID
+                let mut decided_value = None;
+                let pre_votes = match self.get_pre_votes_by_round(self.r) {
+                    Some(v) => v,
+                    None => {
+                        debug!("no pre-votes for this round: {}", self.r);
+                        return Ok(());
+                    }
+                };
+                // Collect n − t valid and properly justified round-r pre-vote messages.
+                if pre_votes.len() == self.threshold() {
+                    // How many votes are 0?
+                    let zero_count = pre_votes
+                        .iter()
+                        .filter(|(_, a)| a.value == PreVoteValue::Zero)
+                        .count();
+
+                    // How many votes are 1?
+                    let one_count = pre_votes
+                        .iter()
+                        .filter(|(_, a)| a.value == PreVoteValue::One)
+                        .count();
+
+                    let (value, just) = if zero_count == self.threshold() {
+                        decided_value = Some(false);
+                        // If there are n − t pre-votes for 0:
+                        //   - value: 0
+                        //   - justification: combination of all pre-votes S-Signature shares
+                        let sig_share: HashMap<&NodeId, &SignatureShare> =
+                            pre_votes.iter().map(|(n, a)| (n, &a.sig_share)).collect();
+                        let sig = self.pub_key_set.combine_signatures(sig_share)?;
+
+                        (
+                            MainVoteValue::Zero,
+                            MainVoteJustification::NoAbstainJustification(sig),
+                        )
+                    } else if one_count == self.threshold() {
+                        decided_value = Some(true);
+                        // If there are n − t pre-votes for 1:
+                        //   - value: 1
+                        //   - justification: combination of all pre-votes S-Signature shares
+                        let sig_share: HashMap<&NodeId, &SignatureShare> =
+                            pre_votes.iter().map(|(n, a)| (n, &a.sig_share)).collect();
+                        let sig = self.pub_key_set.combine_signatures(sig_share)?;
+
+                        (
+                            MainVoteValue::One,
+                            MainVoteJustification::NoAbstainJustification(sig),
+                        )
+                    } else {
```

Comment:
I think this should be an `else if`.

The two branches above check for equality against threshold. Consider what would happen if we only received a `one` pre-vote (no `zero` pre-vote yet). If threshold was 2, the branches above would fail and so it would fall into this case here. But we only want to enter this case when we have pre-votes for zero and one.

Consider changing this `else` to something like the following:
```suggestion
                    } else if let (Some(one_vote), Some(zero_vote)) = (pre_votes.iter().find(|(_,v)| v.value == Zero), pre_votes.iter().find(|(_,v)| v.value == One)) {
```

This has the added benefit of removing the unwraps in this branch